[DPE-3253] Charmed Spark tutorial (#87)
canonical/spark-client-snap: docs/tutorial/t-spark-streaming.md

> ## Streaming workloads with Charmed Spark

Reviewer: This is a really long page that is easy to get lost in - it would be helpful to break it down into sections and list them at the beginning to prepare the user.

Here are the sections I identified in the story:
- Set up new workspace
- Set up Kafka and Zookeper 
- Generate sample streaming data 
- Get credentials for Kafka
- Write a Spark job in Python
- View results

> Spark comes with a built-in support for streaming workloads and Charmed Spark takes it a step further by making it easy to integrate with streaming platforms like Kafka.

Reviewer 2: ```suggestion
Spark comes with a built-in support for streaming workloads via Spark Streaming and Charmed Spark takes it a step further by making it easy to integrate with Kafka, using juju. Kafka is an distributed event-store with a producer/consumer, designed to achieve large throughput while allowing horizontal scalability and high-availability. For more information about Kafka, please refer to [here](https://kafka.apache.org/), and about Spark Streaming, please refer [here](https://spark.apache.org/docs/latest/streaming-programming-guide.html). 
```

Reviewer: Small rewording of @deusebio's suggestion:

Spark comes with a built-in support for streaming workloads via Spark Streaming. Charmed Spark takes it a step further by making it easy to integrate with Kafka using Juju. Kafka is ...

> In this section, we are going to generate some streaming data in Kafka, and then process this stream of data in Spark. We are going to use `juju` to deploy a Kafka setup, where a Juju charm will generate events to Kafka, and a Spark job will consume these in real time to process it.

Reviewer 2: ```suggestion
In this section, we are going to generate some streaming data, push it to Kafka, and then consume the stream of data using Spark, producing some aggregation. We are going to use `juju` to deploy both a Kafka cluster as well as a simple Juju charm will generate and push events to Kafka. We will then show how to setup a Spark job to consume events from Kafka in real time and produce some statistics.
```

First of all, let's start by creating a fresh `juju` model to be used as an experimental workspace.

```bash
juju add-model spark-streaming
```

When you add a Juju model, a Kubernetes namespace of the same name is created automatically. You can verify that by running `kubectl get namespaces` and there you should see a namespace named `spark-streaming`.

The service account `spark` we created in the earlier section is in the `spark` namespace. Let's create a similar service account but now in the `spark-streaming` namespace. We can copy the existing config options from the old service account to the new service account with an addition of `spark.jars.ivy` configuration.

```bash
# Get config from old service account and store in a file
spark-client.service-account-registry get-config \
  --username spark --namespace spark > properties.conf

# Add an extra configuration for the path to cache dependency jars

> echo "spark.jars.ivy=/tmp" >> properties.conf

Reviewer 2: @welpaolo can we skip this with the next release of the snap?

Author: I think so. Removing it.

# Create a new service account and load configurations from the file
spark-client.service-account-registry create \
  --username spark --namespace spark-streaming \
  --properties-file properties.conf
```

> Now, let's create a minimal Kafka and Zookeeper setup. This can be done both easily and quickly using [`zookeeper-k8s`](https://github.com/canonical/zookeeper-k8s-operator) and [`kafka-k8s`](https://charmhub.io/kafka-k8s) charms. Single units of these charms running in our model should be enough.

Reviewer 2: ```suggestion
Now, let's create a minimal Kafka and Zookeeper setup. This can be done both easily and quickly using [`zookeeper-k8s`](https://github.com/canonical/zookeeper-k8s-operator) and [`kafka-k8s`](https://charmhub.io/kafka-k8s) charms. Although this setup is not highly-available, using single units for both should be enough to understand the underlying concepts.
```

zookeeper-k8s           active       1  zookeeper-k8s  3/edge    42  10.152.183.87   no       

Unit              Workload  Agent  Address      Ports  Message
kafka-k8s/0*      blocked   idle   10.1.29.184         missing required zookeeper relation
zookeeper-k8s/0*  active    idle   10.1.29.182         
```

The `kafka-k8s/0` unit is blocked because we have not integrated Kafka with Zookeeper yet. We can do that using:

```bash
juju integrate kafka-k8s zookeeper-k8s
```

After the integration, `juju status` should have an output similar to the following:

```
Model            Controller      Cloud/Region        Version  SLA          Timestamp
spark-streaming  spark-tutorial  microk8s/localhost  3.1.7    unsupported  10:13:55Z

App            Version  Status  Scale  Charm          Channel  Rev  Address         Exposed  Message
kafka-k8s               active      1  kafka-k8s      3/edge    47  10.152.183.242  no       
zookeeper-k8s           active      1  zookeeper-k8s  3/edge    42  10.152.183.87   no       

Unit              Workload  Agent  Address      Ports  Message
kafka-k8s/0*      active    idle   10.1.29.184         
zookeeper-k8s/0*  active    idle   10.1.29.182 
```

> As you can see, both Kafka and Zookeeper charms are in "active" status. However, it can take some amount of time before the app and units are finally transitioned to active state.

Reviewer 2: ```suggestion
As you can see, both Kafka and Zookeeper charms are in "active" status. However, it can take some amount of time before the app and units are finally transitioned to active/idle state.
```

> For us to experiment with the streaming feature, we would want some sample streaming data to be generated in Kafka continuously in real time. For that, we can use the `kafka-test-app` charm as a producer of events. Let's deploy the charm, and integrate it with `kafka-k8s` so that it is able to write messages to Kafka.
> 

Reviewer: ```suggestion
For us to experiment with the streaming feature, we would want some sample streaming data to be generated in Kafka continuously in real time. For that, we can use the `kafka-test-app` charm as a producer of events. 

Let's deploy the charm, and integrate it with `kafka-k8s` so that it is able to write messages to Kafka.
```

```bash

> juju deploy kafka-test-app --series=jammy --channel=edge --config role=producer --config topic_name=spark-streaming-store --config num_messages=100000

Reviewer 2: what about using 3 units, such that we have at least 3 producer that should show some more interesting statistics below?

Author: That's a good idea. Updated.

  username: relation-9
  zookeeper-uris: zookeeper-k8s-0.zookeeper-k8s-endpoints:2181/kafka-k8s
ok: "True"
```

As you can see, the endpoint, username and password to be used to authenticate with Kafka are displayed next to "endpoints", "username" and "password" respectively. It would be easier for us to store the username and password as variables so that they can be used later in the tutorial. For that, we can specify `--format=json` when running the `get-credentials` action, and then filter out username and password using `jq`.

```bash
KAFKA_USERNAME=$(juju run data-integrator/0 get-credentials --format=json | jq -r '.["data-integrator/0"].results.kafka.username')
KAFKA_PASSWORD=$(juju run data-integrator/0 get-credentials --format=json | jq -r '.["data-integrator/0"].results.kafka.password')
KAFKA_ENDPOINT=$(juju run data-integrator/0 get-credentials --format=json | jq -r '.["data-integrator/0"].results.kafka.endpoints')
```

For a moment, let's see the format of an event generated by `kafka-test-app`.

```json
{"timestamp": 1707128717.277745, "_id": "339e0f352b4843bfa0d512b796019a92", "origin": "kafka-test-app-0 (10.1.29.185)", "content": "Message #121"}
```

> 
> As we can see, the value of the "origin" key is the name and the IP address of the unit producing the events. We'll now write a Spark job in Python that counts the number of events grouped by the "origin" key in real time. 
> 
> First of all, we need to authenticate with Kafka and load the events from the `spark-streaming-store` topic . This can be done using `spark.readStream` function as:
> 

Reviewer: When I got here I was a bit confused about how to follow this section - I started copying these python lines into a blank file but then realized that it was more of an explanation/overview of what the final python script would do. So I've reworded it in a way that I think makes it a bit more clear how to follow.
```suggestion

As we can see, the value of the "origin" key is the name and the IP address of the unit producing the events. 

Now we will write a Spark job in Python that counts the number of events grouped by the "origin" key in real time. We will first summarize the structure our script will take, and then provide the full python file for you to copy to your machine.

First, we will authenticate with Kafka and load the events from the `spark-streaming-store` topic . This can be done using the `spark.readStream` function as follows:
```

```bash
aws s3 cp spark_streaming.py s3://spark-tutorial/spark_streaming.py
```

Once the file is copied to S3 server, let's submit a new job to Spark cluster using `spark-submit`. Please note that we need to specify a few extra packages because they are not included by default in Charmed Spark image.

```bash
spark-client.spark-submit \
    --username spark --namespace spark-streaming \
    --deploy-mode cluster \
    --packages org.apache.spark:spark-streaming-kafka-0-10_2.12:3.4.1,org.apache.spark:spark-sql-kafka-0-10_2.12:3.4.1 \
    s3a://spark-tutorial/spark_streaming.py \
        --kafka-endpoint $KAFKA_ENDPOINT \
        --kafka-username $KAFKA_USERNAME \
        --kafka-password $KAFKA_PASSWORD \
        --kafka-topic "spark-streaming-store"
```

The job submission should start a driver pod, which in turn will create executor pods. The executor pods will then transition to the "Running" state and the streaming data is processed continuously by the executor pods.

We can view the status of the pods with the following command in a new shell:

```python
watch -n1 "kubectl get pods -n spark-streaming | grep 'spark-streaming-.*-driver' "
```

> The results are that are written to the console in the program we wrote earlier are actually written to the pod logs. To fetch the pod logs, we first need to know the name of the driver pod. Let's find the name of the driver pod and then fetch pod logs as:

Reviewer 2: ```suggestion
The streaming output - directed to the console - are there being written to the pod logs. To fetch the pod logs, we first need to know the name of the driver pod. Let's find its name to then fetch the logs as:
```

```bash
pod_name=$(kubectl get pods -n spark-streaming | grep "spark-streaming-.*-driver" | tail -n 1 | cut -d' ' -f1)

kubectl logs -n spark-streaming -f $pod_name | grep "Batch: " -A 10 # filter out line starting with "Batch: " and next 10 lines after that line
```

> The option `-f` will tail the pod logs until `Ctrl + C` keys are pressed. If you observe carefully, you can see that new logs are appended every ten seconds, including the results containing the number of events grouped by the origin similar to the following:

Reviewer 2: ```suggestion
The option `-f` will tail the pod logs until `Ctrl + C` keys are pressed. If you observe carefully, you can see that new logs are appended every roughly ten seconds, including the results containing the number of events grouped by the origin, similar to the following:
```

```
...
2024-02-05T16:04:05.189Z [entrypoint] -------------------------------------------
2024-02-05T16:04:05.189Z [entrypoint] Batch: 13
2024-02-05T16:04:05.189Z [entrypoint] -------------------------------------------
2024-02-05T16:04:05.228Z [entrypoint] +--------------------+-----+
2024-02-05T16:04:05.228Z [entrypoint] |              origin|count|
2024-02-05T16:04:05.228Z [entrypoint] +--------------------+-----+
2024-02-05T16:04:05.228Z [entrypoint] |kafka-test-app-0 ...|  272|
2024-02-05T16:04:05.228Z [entrypoint] +--------------------+-----+
2024-02-05T16:04:05.228Z [entrypoint] 
...
```

Bravo! We were successful in processing streaming data with Charmed Spark solution.

> In the next section, we will learn to monitor the status of the job using Spark History Server and the Canonical Observability Stack.

Reviewer 2: ```suggestion
In the next section, we will learn how to monitor the status of the job using Spark History Server and the Canonical Observability Stack.
```
